cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java

```diff
@@ -0,0 +1,228 @@
+import org.apache.hadoop.hbase.HTableDescriptor;
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.HTableInterface;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.coprocessor.BaseRegionObserver;
+import org.apache.hadoop.hbase.coprocessor.ObserverContext;
+import org.apache.hadoop.hbase.coprocessor.RegionCoprocessorEnvironment;
+import org.apache.hadoop.hbase.filter.FilterBase;
+import org.apache.hadoop.hbase.regionserver.InternalScanner;
+import org.apache.hadoop.hbase.regionserver.KeyValueScanner;
+import org.apache.hadoop.hbase.regionserver.ScanType;
+import org.apache.hadoop.hbase.regionserver.Store;
+import org.apache.hadoop.hbase.regionserver.StoreScanner;
+import org.apache.hadoop.hbase.regionserver.compactions.CompactionRequest;
+import org.apache.tephra.coprocessor.TransactionStateCache;
+import org.apache.tephra.persist.TransactionVisibilityState;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * RegionObserver for the Message Table of Transactional Messaging Service.
+ *
+ * This region observer does row eviction during flush time and compact time by using
+ *
+ * i) Time-To-Live (TTL) information in MetadataTable to determine if a message has expired its TTL.
+ * ii) Generation information in MetadataTable to determine if the message belongs to older generation.
+ * iii) Mark data entries with invalidated transactions (loaded from tx.snapshot) so that invalid list can be pruned.
+ *
+ */
+public class MessageTableRegionObserver extends BaseRegionObserver {
+
+  private static final Log LOG = LogFactory.getLog(MessageTableRegionObserver.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_TYPE = new TypeToken<SortedMap<String, String>>() { }.getType();
+
+  private static final byte[] COL_FAMILY = MessagingUtils.Constants.COLUMN_FAMILY;
+  private static final byte[] COL = MessagingUtils.Constants.METADATA_COLUMN;
+
+  private static int prefixLength;
+  private static LoadingCache<ByteBuffer, Map<String, String>> topicCache;
+  private static TransactionStateCache txStateCache;
+
+  private HTableInterface metadataTable;
+  private CConfigurationReader cConfReader;
+
+  @Override
+  public void start(CoprocessorEnvironment env) throws IOException {
+    if (env instanceof RegionCoprocessorEnvironment) {
+      HTableDescriptor tableDesc = ((RegionCoprocessorEnvironment) env).getRegion().getTableDesc();
+      String metadataTableNamespace = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAMESPACE);
+      String metadataTableName = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAME);
```

Comment:
I have modified this so that we can read the metadataTableName from cConf. However, it is not straightforward to remove metadataTableNamespace attribute. This is because we use tableUtil to convert the cdap namespace to hbase namespace and then we write to the data table as an attribute. We can also get HBaseTableUtil in the coprocessor and do the conversion (and it will work since for system namespace we don't look up NamespaceQueryAdmin) but we can avoid it if we use the attribute.